// @runloop/api-client/resources/repositories.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../resource.mjs";
import { isRequestOptions } from "../core.mjs";
import { RepositoriesCursorIDPage } from "../pagination.mjs";
export class Repositories extends APIResource {
    /**
     * Create a connection to a Github Repository and trigger an initial inspection of
     * the repo's technical stack and developer environment requirements.
     */
    create(body, options) {
        return this._client.post('/v1/repositories', { body, ...options });
    }
    /**
     * Get Repository Connection details including latest inspection status and
     * generated respository insights.
     */
    retrieve(id, options) {
        return this._client.get(`/v1/repositories/${id}`, options);
    }
    list(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/v1/repositories', RepositoryConnectionViewsRepositoriesCursorIDPage, {
            query,
            ...options,
        });
    }
    /**
     * Permanently Delete a Repository Connection including any automatically generated
     * inspection insights.
     */
    delete(id, body, options) {
        return this._client.post(`/v1/repositories/${id}/delete`, { body, ...options });
    }
    /**
     * List all analyzed versions of a repository connection including automatically
     * generated insights for each version.
     */
    versions(id, options) {
        return this._client.get(`/v1/repositories/${id}/versions`, options);
    }
}
export class RepositoryConnectionViewsRepositoriesCursorIDPage extends RepositoriesCursorIDPage {
}
Repositories.RepositoryConnectionViewsRepositoriesCursorIDPage =
    RepositoryConnectionViewsRepositoriesCursorIDPage;
