// @runloop/api-client/resources/blueprints.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../resource.mjs";
import { isRequestOptions } from "../core.mjs";
import { BlueprintsCursorIDPage } from "../pagination.mjs";
export class Blueprints extends APIResource {
    /**
     * Starts build of custom defined container Blueprint. The Blueprint will begin in
     * the 'provisioning' step and transition to the 'building' step once it is
     * selected off the build queue., Upon build complete it will transition to
     * 'building_complete' if the build is successful.
     */
    create(body, options) {
        return this._client.post('/v1/blueprints', { body, ...options });
    }
    /**
     * Get the details of a previously created Blueprint including the build status.
     */
    retrieve(id, options) {
        return this._client.get(`/v1/blueprints/${id}`, options);
    }
    list(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/v1/blueprints', BlueprintViewsBlueprintsCursorIDPage, {
            query,
            ...options,
        });
    }
    /**
     * Get all logs from the building of a Blueprint.
     */
    logs(id, options) {
        return this._client.get(`/v1/blueprints/${id}/logs`, options);
    }
    /**
     * Preview building a Blueprint with the specified configuration. You can take the
     * resulting Dockerfile and test out your build using any local docker tooling.
     */
    preview(body, options) {
        return this._client.post('/v1/blueprints/preview', { body, ...options });
    }
}
export class BlueprintViewsBlueprintsCursorIDPage extends BlueprintsCursorIDPage {
}
Blueprints.BlueprintViewsBlueprintsCursorIDPage = BlueprintViewsBlueprintsCursorIDPage;
