// @runloop/api-client/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnprocessableEntityError = exports.PermissionDeniedError = exports.InternalServerError = exports.AuthenticationError = exports.BadRequestError = exports.RateLimitError = exports.ConflictError = exports.NotFoundError = exports.APIUserAbortError = exports.APIConnectionTimeoutError = exports.APIConnectionError = exports.APIError = exports.RunloopError = exports.fileFromPath = exports.toFile = exports.Runloop = void 0;
const Core = __importStar(require("./core.js"));
const Errors = __importStar(require("./error.js"));
const Pagination = __importStar(require("./pagination.js"));
const Uploads = __importStar(require("./uploads.js"));
const API = __importStar(require("./resources/index.js"));
const blueprints_1 = require("./resources/blueprints.js");
const repositories_1 = require("./resources/repositories.js");
const devboxes_1 = require("./resources/devboxes/devboxes.js");
/**
 * API Client for interfacing with the Runloop API.
 */
class Runloop extends Core.APIClient {
    /**
     * API Client for interfacing with the Runloop API.
     *
     * @param {string | undefined} [opts.bearerToken=process.env['RUNLOOP_API_KEY'] ?? undefined]
     * @param {string} [opts.baseURL=process.env['RUNLOOP_BASE_URL'] ?? https://api.runloop.ai] - Override the default base URL for the API.
     * @param {number} [opts.timeout=1 minute] - The maximum amount of time (in milliseconds) the client will wait for a response before timing out.
     * @param {number} [opts.httpAgent] - An HTTP agent used to manage HTTP(s) connections.
     * @param {Core.Fetch} [opts.fetch] - Specify a custom `fetch` function implementation.
     * @param {number} [opts.maxRetries=2] - The maximum number of times the client will retry a request.
     * @param {Core.Headers} opts.defaultHeaders - Default headers to include with every request to the API.
     * @param {Core.DefaultQuery} opts.defaultQuery - Default query parameters to include with every request to the API.
     */
    constructor({ baseURL = Core.readEnv('RUNLOOP_BASE_URL'), bearerToken = Core.readEnv('RUNLOOP_API_KEY'), ...opts } = {}) {
        if (bearerToken === undefined) {
            throw new Errors.RunloopError("The RUNLOOP_API_KEY environment variable is missing or empty; either provide it, or instantiate the Runloop client with an bearerToken option, like new Runloop({ bearerToken: 'My Bearer Token' }).");
        }
        const options = {
            bearerToken,
            ...opts,
            baseURL: baseURL || `https://api.runloop.ai`,
        };
        super({
            baseURL: options.baseURL,
            timeout: options.timeout ?? 60000 /* 1 minute */,
            httpAgent: options.httpAgent,
            maxRetries: options.maxRetries,
            fetch: options.fetch,
        });
        this.blueprints = new API.Blueprints(this);
        this.devboxes = new API.Devboxes(this);
        this.repositories = new API.Repositories(this);
        this._options = options;
        this.idempotencyHeader = 'x-request-id';
        this.bearerToken = bearerToken;
    }
    defaultQuery() {
        return this._options.defaultQuery;
    }
    defaultHeaders(opts) {
        return {
            ...super.defaultHeaders(opts),
            ...this._options.defaultHeaders,
        };
    }
    authHeaders(opts) {
        return { Authorization: `Bearer ${this.bearerToken}` };
    }
}
exports.Runloop = Runloop;
_a = Runloop;
Runloop.Runloop = _a;
Runloop.DEFAULT_TIMEOUT = 60000; // 1 minute
Runloop.RunloopError = Errors.RunloopError;
Runloop.APIError = Errors.APIError;
Runloop.APIConnectionError = Errors.APIConnectionError;
Runloop.APIConnectionTimeoutError = Errors.APIConnectionTimeoutError;
Runloop.APIUserAbortError = Errors.APIUserAbortError;
Runloop.NotFoundError = Errors.NotFoundError;
Runloop.ConflictError = Errors.ConflictError;
Runloop.RateLimitError = Errors.RateLimitError;
Runloop.BadRequestError = Errors.BadRequestError;
Runloop.AuthenticationError = Errors.AuthenticationError;
Runloop.InternalServerError = Errors.InternalServerError;
Runloop.PermissionDeniedError = Errors.PermissionDeniedError;
Runloop.UnprocessableEntityError = Errors.UnprocessableEntityError;
Runloop.toFile = Uploads.toFile;
Runloop.fileFromPath = Uploads.fileFromPath;
Runloop.Blueprints = blueprints_1.Blueprints;
Runloop.BlueprintViewsBlueprintsCursorIDPage = blueprints_1.BlueprintViewsBlueprintsCursorIDPage;
Runloop.Devboxes = devboxes_1.Devboxes;
Runloop.DevboxViewsDevboxesCursorIDPage = devboxes_1.DevboxViewsDevboxesCursorIDPage;
Runloop.DevboxSnapshotViewsDiskSnapshotsCursorIDPage = devboxes_1.DevboxSnapshotViewsDiskSnapshotsCursorIDPage;
Runloop.Repositories = repositories_1.Repositories;
Runloop.RepositoryConnectionViewsRepositoriesCursorIDPage = repositories_1.RepositoryConnectionViewsRepositoriesCursorIDPage;
var uploads_1 = require("./uploads.js");
Object.defineProperty(exports, "toFile", { enumerable: true, get: function () { return uploads_1.toFile; } });
Object.defineProperty(exports, "fileFromPath", { enumerable: true, get: function () { return uploads_1.fileFromPath; } });
var error_1 = require("./error.js");
Object.defineProperty(exports, "RunloopError", { enumerable: true, get: function () { return error_1.RunloopError; } });
Object.defineProperty(exports, "APIError", { enumerable: true, get: function () { return error_1.APIError; } });
Object.defineProperty(exports, "APIConnectionError", { enumerable: true, get: function () { return error_1.APIConnectionError; } });
Object.defineProperty(exports, "APIConnectionTimeoutError", { enumerable: true, get: function () { return error_1.APIConnectionTimeoutError; } });
Object.defineProperty(exports, "APIUserAbortError", { enumerable: true, get: function () { return error_1.APIUserAbortError; } });
Object.defineProperty(exports, "NotFoundError", { enumerable: true, get: function () { return error_1.NotFoundError; } });
Object.defineProperty(exports, "ConflictError", { enumerable: true, get: function () { return error_1.ConflictError; } });
Object.defineProperty(exports, "RateLimitError", { enumerable: true, get: function () { return error_1.RateLimitError; } });
Object.defineProperty(exports, "BadRequestError", { enumerable: true, get: function () { return error_1.BadRequestError; } });
Object.defineProperty(exports, "AuthenticationError", { enumerable: true, get: function () { return error_1.AuthenticationError; } });
Object.defineProperty(exports, "InternalServerError", { enumerable: true, get: function () { return error_1.InternalServerError; } });
Object.defineProperty(exports, "PermissionDeniedError", { enumerable: true, get: function () { return error_1.PermissionDeniedError; } });
Object.defineProperty(exports, "UnprocessableEntityError", { enumerable: true, get: function () { return error_1.UnprocessableEntityError; } });
exports = module.exports = Runloop;
exports.default = Runloop;
