// @runloop/api-client/resources/blueprints.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlueprintViewsBlueprintsCursorIDPage = exports.Blueprints = void 0;
const resource_1 = require("../resource.js");
const core_1 = require("../core.js");
const pagination_1 = require("../pagination.js");
class Blueprints extends resource_1.APIResource {
    /**
     * Starts build of custom defined container Blueprint. The Blueprint will begin in
     * the 'provisioning' step and transition to the 'building' step once it is
     * selected off the build queue., Upon build complete it will transition to
     * 'building_complete' if the build is successful.
     */
    create(body, options) {
        return this._client.post('/v1/blueprints', { body, ...options });
    }
    /**
     * Get the details of a previously created Blueprint including the build status.
     */
    retrieve(id, options) {
        return this._client.get(`/v1/blueprints/${id}`, options);
    }
    list(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/v1/blueprints', BlueprintViewsBlueprintsCursorIDPage, {
            query,
            ...options,
        });
    }
    /**
     * Get all logs from the building of a Blueprint.
     */
    logs(id, options) {
        return this._client.get(`/v1/blueprints/${id}/logs`, options);
    }
    /**
     * Preview building a Blueprint with the specified configuration. You can take the
     * resulting Dockerfile and test out your build using any local docker tooling.
     */
    preview(body, options) {
        return this._client.post('/v1/blueprints/preview', { body, ...options });
    }
}
exports.Blueprints = Blueprints;
class BlueprintViewsBlueprintsCursorIDPage extends pagination_1.BlueprintsCursorIDPage {
}
exports.BlueprintViewsBlueprintsCursorIDPage = BlueprintViewsBlueprintsCursorIDPage;
Blueprints.BlueprintViewsBlueprintsCursorIDPage = BlueprintViewsBlueprintsCursorIDPage;
